py/src/braintrust/logger.py

```diff
@@ -1480,6 +1547,134 @@ def __exit__(self, type, value, callback):
+        :param scores: A dictionary of numeric values (between 0 and 1) to log. The scores should give you a variety of signals that help you determine how accurate the outputs are compared to what you expect and diagnose failures. For example, a summarization app might have one score that tells you how accurate the summary is, and another that measures the word similarity between the generated and grouth truth summary. The word similarity score could help you determine whether the summarization was covering similar concepts or not. You can use these scores to help you sort, filter, and compare logs.
+        :param metadata: (Optional) a dictionary with additional data about the test example, model outputs, or just about anything else that's relevant, that you can use to help find and analyze examples later. For example, you could log the `prompt`, example's `id`, or anything else that would be useful to slice/dice later. The values in `metadata` can be any JSON-serializable type, but its keys must be strings.
+        :param metrics: (Optional) a dictionary of metrics to log. The following keys are populated automatically and should not be specified: "start", "end", "caller_functionname", "caller_filename", "caller_lineno".
+        :param id: (Optional) a unique identifier for the event. If you don't provide one, BrainTrust will generate one for you.
+        """
+        span = self.start_span(
+            start_time=self.last_start_time,
+            input=input,
+            output=output,
+            expected=expected,
+            scores=scores,
+            metadata=metadata,
+            metrics=metrics,
+            id=id,
+        )
+        self.last_start_time = span.end()
+
+        if not self.async_flush:
+            self.logger.flush()
+
+        return span.id
+
+    def _perform_lazy_login(self):
+        if not self._logged_in:
+            self._lazy_login()
+            self.last_start_time = time.time()
+            self._logged_in = True
+
+    def start_span(self, name="root", span_attributes={}, start_time=None, set_current=None, **event):
+        """Create a new toplevel span. The name parameter is optional and defaults to "root".
+
+        See `Span.start_span` for full details
+        """
+        self._perform_lazy_login()
+        return SpanImpl(
+            bg_logger=self.logger,
+            name=name,
+            span_attributes=span_attributes,
+            start_time=start_time,
+            set_current=set_current,
+            event=event,
+            root_project=self.project,
+        )
+
+    def __enter__(self):
+        self._perform_lazy_login()
+        if self.set_current:
+            self._context_token = _state.current_logger.set(self)
+        return self
+
+    def __exit__(self, type, value, callback):
+        del type, value, callback
+
+        if self.set_current:
+            _state.current_logger.reset(self._context_token)
```

Comment:
I think `exit` must also call flush. When I ran the `fake_experiments` benchmark against my local chalice without this, I'd always get a "broken pipe" error at the end of the benchmark, and not all the logs would get written.

I think the reason is that the benchmark uses `multiprocessing` and for some reason the `atexit` handlers are not called when subprocesses exit 🤷

Reply:
can do